File: Java/src/_0028_Implement_strStr/Solution.java
package _0028_Implement_strStr;

/**
 * Date: 2020/9/19 21:22
 * Content: kmp算法
 * 执行用时：3 ms, 在所有 Java 提交中击败了39.16%的用户
 * 内存消耗：38.4 MB, 在所有 Java 提交中击败了38.65%的用户
 */
public class Solution {

    public int strStr(String haystack, String needle) {

        if (needle == null || needle.length() == 0) return 0;
        if (needle.length() > haystack.length()) return -1;

        int m = haystack.length(), n = needle.length();
        char[] s = (" " + haystack).toCharArray(), p = (" " + needle).toCharArray();

        int[] ne = new int[n + 10];
        for (int i = 2, j = 0; i <= n; i++) {
            while (j != 0 && p[i] != p[j + 1]) j = ne[j];
            if (p[i] == p[j + 1]) j++;
            ne[i] = j;
        }

        for (int i = 1, j = 0; i <= m; i++) {
            while (j != 0 && s[i] != p[j + 1]) j = ne[j];
            if (s[i] == p[j + 1]) j++;
            if (j == n) return i - n;
        }
        return -1;
    }

    public static void main(String[] args) {

        System.out.println((new Solution()).strStr("", ""));  // 0
        System.out.println((new Solution()).strStr("hello", "ll"));  // 2
        System.out.println((new Solution()).strStr("aaaaa", "bba"));  // -1
        System.out.println((new Solution()).strStr("mississippi", "issippi"));  // 4
    }
}

File: Python/_0028_Implement_strStr/Solution.py
# coding=utf-8


# 执行用时：60 ms, 在所有 Python3 提交中击败了16.16%的用户
# 内存消耗：16.4 MB, 在所有 Python3 提交中击败了5.06%的用户KMP
class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        if len(needle) == 0:
            return 0
        if len(needle) > len(haystack):
            return -1

        n = len(needle); m = len(haystack)
        p = " " + needle; s = " " + haystack
        ne = [0 for _ in range(n + 10)]

        j = 0
        for i in range(2, n + 1):
            while j > 0 and p[i] != p[j + 1]:
                j = ne[j]
            if p[i] == p[j + 1]:
                j += 1
            ne[i] = j

        j = 0
        for i in range(1, m + 1):
            while j > 0 and s[i] != p[j + 1]:
                j = ne[j]
            if s[i] == p[j + 1]:
                j += 1
            if j == n:
                return i - n
        return -1


if __name__ == "__main__":
    print(Solution().strStr("", ""))  # 0
    print(Solution().strStr("hello", "ll"))  # 2
    print(Solution().strStr("aaaaa", "bba"))  # -1
    print(Solution().strStr("mississippi", "issippi"))  # 4
